### PR TITLE
fix: Run controller as un-privileged

### PR DESCRIPTION
**Pull request by sbose78**
possibly part of https://github.com/argoproj/argoproj/issues/40

Checklist:

* [x] My organization is added to [USERS.md](https://github.com/argoproj/argo-workflows/blob/master/USERS.md).

### CODE REVIEW

**Comment by sbose78** [comment]
Yet to get `make codegen` working :(  Wanted to put it out here to see if anything fails. Question, are these tests run against a Kubernetes cluster?

**Comment by simster7** [comment]
@sbose78 What is the issue with `make codegen`?

File: manifests/base/argo-server/argo-server-deployment.yaml [hunk 1]
```diff
@@ -16,6 +16,9 @@ spec:
         - name: argo-server
           image: argoproj/argocli:latest
           securityContext:
+            readOnlyRootFilesystem: true
```
Comment (alexec) [inline manifests/base/argo-server/argo-server-deployment.yaml:19]: The server needs to write to /tmp for artefact download.

File: manifests/base/workflow-controller/workflow-controller-deployment.yaml [hunk 2]
```diff
@@ -16,6 +16,9 @@ spec:
         - name: workflow-controller
           image: argoproj/workflow-controller:latest
           securityContext:
```
Comment (alexec) [inline manifests/base/workflow-controller/workflow-controller-deployment.yaml:18]: Any reason not to merge with the pod security context?

File: manifests/base/argo-server/argo-server-deployment.yaml (hunk 1 above)
Comment (sbose78) [reply manifests/base/argo-server/argo-server-deployment.yaml:19]: Thanks, I do see the following present already, would that suffice @alexec ?
```
  volumeMounts:
            - mountPath: /tmp
              name: tmp
```

File: manifests/base/workflow-controller/workflow-controller-deployment.yaml (hunk 2 above)
Comment (sbose78) [reply manifests/base/workflow-controller/workflow-controller-deployment.yaml:18]: Theoretically, the pod may have other containers which may not function with those attributes enabled. However, I'll be happy to move this to the pod level if you suggest so.

Comment (alexec) [reply manifests/base/workflow-controller/workflow-controller-deployment.yaml:18]: Injected side-cars? Istio? I think I'd like it all in one place. I don't really mind where that is.

File: manifests/base/argo-server/argo-server-deployment.yaml (hunk 1 above)
Comment (alexec) [reply manifests/base/argo-server/argo-server-deployment.yaml:19]: Have you checked this works?

**Comment by sbose78** [comment]
I'll make a separate PR for `server` - limiting this one for controller only.

I seem to be having the `time="2021-03-22T04:07:09.974Z" level=fatal msg="open argo-server.key: permission denied"` issue on OpenShift with the manifests from master. 

Related: https://github.com/argoproj/argo-workflows/pull/4750

**Review by alexec — APPROVED** [review]

**Review by alexec — APPROVED** [review]

**Pull request merged**